File: backend/src/services/workspaceService.ts
import { query, transaction } from '../db/client.js';
import { Role } from '../models/types.js';
import { generateId } from '../utils/id.js';

export interface CreateWorkspaceData {
  name: string;
  description?: string;
  logo?: string;
  ownerId: string;
}

export interface UpdateWorkspaceData {
  name?: string;
  description?: string;
  logo?: string;
}

export interface WorkspaceWithMembers {
  id: string;
  name: string;
  description: string | null;
  logo: string | null;
  ownerId: string;
  createdAt: Date;
  updatedAt: Date;
  members: Array<{
    id: string;
    userId: string;
    role: Role;
    joinedAt: Date;
    user: {
      id: string;
      email: string;
      firstName: string | null;
      lastName: string | null;
      avatar: string | null;
    };
  }>;
}

/**
 * Create a new workspace with the owner as the first member
 */
export async function createWorkspace(data: CreateWorkspaceData): Promise<WorkspaceWithMembers> {
  return transaction(async (client) => {
    const workspaceId = generateId();
    const memberId = generateId();
    const now = new Date();

    // Create the workspace
    await client.query(
      `INSERT INTO workspaces (id, name, description, logo, "ownerId", "createdAt", "updatedAt")
       VALUES ($1, $2, $3, $4, $5, $6, $7)`,
      [workspaceId, data.name, data.description || null, data.logo || null, data.ownerId, now, now]
    );

    // Add the owner as a workspace member with OWNER role
    await client.query(
      `INSERT INTO workspace_members (id, "workspaceId", "userId", role, "joinedAt", "updatedAt")
       VALUES ($1, $2, $3, $4, $5, $6)`,
      [memberId, workspaceId, data.ownerId, Role.OWNER, now, now]
    );

    // Get the complete workspace with members using the transaction client
    const result = await client.query(
      `SELECT 
        w.id, w.name, w.description, w.logo, w."ownerId", w."createdAt", w."updatedAt",
        wm.id as member_id, wm."userId", wm.role, wm."joinedAt",
        u.email, u."firstName", u."lastName", u.avatar
       FROM workspaces w
       LEFT JOIN workspace_members wm ON w.id = wm."workspaceId"
       LEFT JOIN users u ON wm."userId" = u.id
       WHERE w.id = $1
       ORDER BY wm."joinedAt" ASC`,
      [workspaceId]
    );

    if (result.rows.length === 0) {
      throw new Error('Failed to create workspace');
    }

    const firstRow = result.rows[0];
    const workspace: WorkspaceWithMembers = {
      id: firstRow.id,
      name: firstRow.name,
      description: firstRow.description,
      logo: firstRow.logo,
      ownerId: firstRow.ownerId,
      createdAt: firstRow.createdAt,
      updatedAt: firstRow.updatedAt,
      members: []
    };

    // Build members array from joined results
    for (const row of result.rows) {
      if (row.member_id) {
        workspace.members.push({
          id: row.member_id,
          userId: row.userId,
          role: row.role as Role,
          joinedAt: row.joinedAt,
          user: {
            id: row.userId,
            email: row.email,
            firstName: row.firstName,
            lastName: row.lastName,
            avatar: row.avatar
          }
        });
      }
    }

    return workspace;
  });
}

/**
 * Get workspace by ID with members
 */
export async function getWorkspaceById(workspaceId: string): Promise<WorkspaceWithMembers | null> {
  const result = await query(
    `SELECT 
      w.id, w.name, w.description, w.logo, w."ownerId", w."createdAt", w."updatedAt",
      wm.id as member_id, wm."userId", wm.role, wm."joinedAt",
      u.email, u."firstName", u."lastName", u.avatar
     FROM workspaces w
     LEFT JOIN workspace_members wm ON w.id = wm."workspaceId"
     LEFT JOIN users u ON wm."userId" = u.id
     WHERE w.id = $1
     ORDER BY wm."joinedAt" ASC`,
    [workspaceId]
  );

  if (result.rows.length === 0) {
    return null;
  }

  const firstRow = result.rows[0];
  const workspace: WorkspaceWithMembers = {
    id: firstRow.id,
    name: firstRow.name,
    description: firstRow.description,
    logo: firstRow.logo,
    ownerId: firstRow.ownerId,
    createdAt: firstRow.createdAt,
    updatedAt: firstRow.updatedAt,
    members: []
  };

  // Build members array from joined results
  for (const row of result.rows) {
    if (row.member_id) {
      workspace.members.push({
        id: row.member_id,
        userId: row.userId,
        role: row.role as Role,
        joinedAt: row.joinedAt,
        user: {
          id: row.userId,
          email: row.email,
          firstName: row.firstName,
          lastName: row.lastName,
          avatar: row.avatar
        }
      });
    }
  }

  return workspace;
}

/**
 * Get all workspaces for a user (where they are a member)
 */
export async function getWorkspacesForUser(userId: string): Promise<WorkspaceWithMembers[]> {
  const result = await query(
    `SELECT 
      w.id, w.name, w.description, w.logo, w."ownerId", w."createdAt", w."updatedAt",
      wm.id as member_id, wm."userId", wm.role, wm."joinedAt",
      u.email, u."firstName", u."lastName", u.avatar
     FROM workspaces w
     INNER JOIN workspace_members wm ON w.id = wm."workspaceId"
     LEFT JOIN workspace_members all_members ON w.id = all_members."workspaceId"
     LEFT JOIN users u ON all_members."userId" = u.id
     WHERE wm."userId" = $1
     ORDER BY w."createdAt" DESC, all_members."joinedAt" ASC`,
    [userId]
  );

  // Group results by workspace
  const workspaceMap = new Map<string, WorkspaceWithMembers>();

  for (const row of result.rows) {
    if (!workspaceMap.has(row.id)) {
      workspaceMap.set(row.id, {
        id: row.id,
        name: row.name,
        description: row.description,
        logo: row.logo,
        ownerId: row.ownerId,
        createdAt: row.createdAt,
        updatedAt: row.updatedAt,
        members: []
      });
    }

    const workspace = workspaceMap.get(row.id)!;

    // Add member if not already added
    if (row.member_id && !workspace.members.find(m => m.id === row.member_id)) {
      workspace.members.push({
        id: row.member_id,
        userId: row.userId,
        role: row.role as Role,
        joinedAt: row.joinedAt,
        user: {
          id: row.userId,
          email: row.email,
          firstName: row.firstName,
          lastName: row.lastName,
          avatar: row.avatar
        }
      });
    }
  }

  return Array.from(workspaceMap.values());
}

/**
 * Update workspace information
 */
export async function updateWorkspace(
  workspaceId: string,
  data: UpdateWorkspaceData
): Promise<WorkspaceWithMembers | null> {
  const updates: string[] = [];
  const values: unknown[] = [];
  let paramIndex = 1;

  if (data.name !== undefined) {
    updates.push(`name = $${paramIndex++}`);
    values.push(data.name);
  }

  if (data.description !== undefined) {
    updates.push(`description = $${paramIndex++}`);
    values.push(data.description);
  }

  if (data.logo !== undefined) {
    updates.push(`logo = $${paramIndex++}`);
    values.push(data.logo);
  }

  if (updates.length === 0) {
    return getWorkspaceById(workspaceId);
  }

  updates.push(`"updatedAt" = $${paramIndex++}`);
  values.push(new Date());
  values.push(workspaceId);

  await query(
    `UPDATE workspaces SET ${updates.join(', ')} WHERE id = $${paramIndex}`,
    values
  );

  return getWorkspaceById(workspaceId);
}

/**
 * Delete a workspace and all related data
 */
export async function deleteWorkspace(workspaceId: string): Promise<boolean> {
  const result = await query(
    'DELETE FROM workspaces WHERE id = $1',
    [workspaceId]
  );

  return (result.rowCount ?? 0) > 0;
}

/**
 * Add a member to a workspace
 */
export async function addWorkspaceMember(
  workspaceId: string,
  userId: string,
  role: Role = Role.MEMBER
): Promise<boolean> {
  try {
    const memberId = generateId();
    const now = new Date();

    await query(
      `INSERT INTO workspace_members (id, "workspaceId", "userId", role, "joinedAt", "updatedAt")
       VALUES ($1, $2, $3, $4, $5, $6)`,
      [memberId, workspaceId, userId, role, now, now]
    );

    return true;
  } catch (error: unknown) {
    // Handle unique constraint violation (user already a member)
    if (error && typeof error === 'object' && 'code' in error && error.code === '23505') {
      return false;
    }
    throw error;
  }
}

/**
 * Remove a member from a workspace
 */
export async function removeWorkspaceMember(
  workspaceId: string,
  userId: string
): Promise<boolean> {
  const result = await query(
    'DELETE FROM workspace_members WHERE "workspaceId" = $1 AND "userId" = $2',
    [workspaceId, userId]
  );

  return (result.rowCount ?? 0) > 0;
}

/**
 * Update a member's role in a workspace
 */
export async function updateWorkspaceMemberRole(
  workspaceId: string,
  userId: string,
  role: Role
): Promise<boolean> {
  const result = await query(
    `UPDATE workspace_members 
     SET role = $1, "updatedAt" = $2
     WHERE "workspaceId" = $3 AND "userId" = $4`,
    [role, new Date(), workspaceId, userId]
  );

  return (result.rowCount ?? 0) > 0;
}

/**
 * Check if a user is a member of a workspace
 */
export async function isWorkspaceMember(
  workspaceId: string,
  userId: string
): Promise<{ isMember: boolean; role?: Role }> {
  const result = await query(
    'SELECT role FROM workspace_members WHERE "workspaceId" = $1 AND "userId" = $2',
    [workspaceId, userId]
  );

  if (result.rows.length === 0) {
    return { isMember: false };
  }

  return {
    isMember: true,
    role: result.rows[0].role as Role
  };
}

/**
 * Get workspace members
 */
export async function getWorkspaceMembers(workspaceId: string) {
  const result = await query(
    `SELECT 
      wm.id, wm."userId", wm.role, wm."joinedAt",
      u.email, u."firstName", u."lastName", u.avatar
     FROM workspace_members wm
     JOIN users u ON wm."userId" = u.id
     WHERE wm."workspaceId" = $1
     ORDER BY wm."joinedAt" ASC`,
    [workspaceId]
  );

  return result.rows.map(row => ({
    id: row.id,
    userId: row.userId,
    role: row.role as Role,
    joinedAt: row.joinedAt,
    user: {
      id: row.userId,
      email: row.email,
      firstName: row.firstName,
      lastName: row.lastName,
      avatar: row.avatar
    }
  }));
}

File: internal/services/workspace_service.go
package services

import (
	"context"
	"fmt"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"

	"project-management-platform/internal/models"
	"project-management-platform/internal/repository"
)

// WorkspaceService handles workspace business logic
type WorkspaceService struct {
	workspaceRepo repository.WorkspaceRepository
	userRepo      repository.UserRepository
	emailService  EmailService
}

// NewWorkspaceService creates a new workspace service
func NewWorkspaceService(workspaceRepo repository.WorkspaceRepository, userRepo repository.UserRepository, emailService EmailService) *WorkspaceService {
	return &WorkspaceService{
		workspaceRepo: workspaceRepo,
		userRepo:      userRepo,
		emailService:  emailService,
	}
}

// CreateWorkspaceRequest represents a workspace creation request
type CreateWorkspaceRequest struct {
	Name        string `json:"name" validate:"required,min=1,max=100"`
	Description string `json:"description,omitempty" validate:"max=500"`
	Logo        string `json:"logo,omitempty"`
}

// UpdateWorkspaceRequest represents a workspace update request
type UpdateWorkspaceRequest struct {
	Name        *string                   `json:"name,omitempty" validate:"omitempty,min=1,max=100"`
	Description *string                   `json:"description,omitempty" validate:"omitempty,max=500"`
	Logo        *string                   `json:"logo,omitempty"`
	Settings    *models.WorkspaceSettings `json:"settings,omitempty"`
}

// WorkspaceResponse represents a workspace response
type WorkspaceResponse struct {
	ID          string                   `json:"id"`
	Name        string                   `json:"name"`
	Description string                   `json:"description,omitempty"`
	Logo        string                   `json:"logo,omitempty"`
	OwnerID     string                   `json:"ownerId"`
	Members     []models.WorkspaceMember `json:"members"`
	Settings    models.WorkspaceSettings `json:"settings"`
	CreatedAt   time.Time                `json:"createdAt"`
	UpdatedAt   time.Time                `json:"updatedAt"`
}

// Create creates a new workspace
func (s *WorkspaceService) Create(ctx context.Context, userID primitive.ObjectID, req CreateWorkspaceRequest) (*WorkspaceResponse, error) {
	// Validate user exists
	_, err := s.userRepo.GetByID(ctx, userID)
	if err != nil {
		return nil, fmt.Errorf("failed to get user: %w", err)
	}

	// Create workspace
	workspace := models.NewWorkspace(req.Name, req.Description, userID)
	if req.Logo != "" {
		workspace.Logo = req.Logo
	}

	result, err := s.workspaceRepo.Create(ctx, workspace)
	if err != nil {
		return nil, fmt.Errorf("failed to create workspace: %w", err)
	}

	// Set the ID from the insert result
	workspace.ID = result.InsertedID.(primitive.ObjectID)

	return s.toWorkspaceResponse(workspace), nil
}

// GetByID retrieves a workspace by ID
func (s *WorkspaceService) GetByID(ctx context.Context, userID, workspaceID primitive.ObjectID) (*WorkspaceResponse, error) {
	workspace, err := s.workspaceRepo.GetByID(ctx, workspaceID)
	if err != nil {
		return nil, fmt.Errorf("failed to get workspace: %w", err)
	}

	// Check if user has access to this workspace
	if !workspace.IsMember(userID) {
		return nil, fmt.Errorf("access denied: user is not a member of this workspace")
	}

	return s.toWorkspaceResponse(workspace), nil
}

// GetByUserID retrieves workspaces where the user is a member
func (s *WorkspaceService) GetByUserID(ctx context.Context, userID primitive.ObjectID) ([]*WorkspaceResponse, error) {
	workspaces, err := s.workspaceRepo.GetByUserID(ctx, userID)
	if err != nil {
		return nil, fmt.Errorf("failed to get workspaces: %w", err)
	}

	responses := make([]*WorkspaceResponse, len(workspaces))
	for i, workspace := range workspaces {
		responses[i] = s.toWorkspaceResponse(workspace)
	}

	return responses, nil
}

// Update updates a workspace
func (s *WorkspaceService) Update(ctx context.Context, userID, workspaceID primitive.ObjectID, req UpdateWorkspaceRequest) (*WorkspaceResponse, error) {
	// Get workspace and check permissions
	workspace, err := s.workspaceRepo.GetByID(ctx, workspaceID)
	if err != nil {
		return nil, fmt.Errorf("failed to get workspace: %w", err)
	}

	// Check if user is admin
	if !workspace.IsAdmin(userID) {
		return nil, fmt.Errorf("access denied: user is not an admin of this workspace")
	}

	// Build update document
	update := bson.M{}

	if req.Name != nil {
		if *req.Name == "" {
			return nil, fmt.Errorf("workspace name cannot be empty")
		}
		update["name"] = *req.Name
	}

	if req.Description != nil {
		update["description"] = *req.Description
	}

	if req.Logo != nil {
		update["logo"] = *req.Logo
	}

	if req.Settings != nil {
		// Validate settings
		if req.Settings.Visibility != "" &&
			req.Settings.Visibility != models.WorkspaceVisibilityPrivate &&
			req.Settings.Visibility != models.WorkspaceVisibilityPublic {
			return nil, fmt.Errorf("invalid workspace visibility: %s", req.Settings.Visibility)
		}
		update["settings"] = *req.Settings
	}

	if len(update) == 0 {
		return nil, fmt.Errorf("no fields to update")
	}

	// Update workspace
	_, err = s.workspaceRepo.Update(ctx, workspaceID, update)
	if err != nil {
		return nil, fmt.Errorf("failed to update workspace: %w", err)
	}

	// Get updated workspace
	updatedWorkspace, err := s.workspaceRepo.GetByID(ctx, workspaceID)
	if err != nil {
		return nil, fmt.Errorf("failed to get updated workspace: %w", err)
	}

	return s.toWorkspaceResponse(updatedWorkspace), nil
}

// Delete deletes a workspace
func (s *WorkspaceService) Delete(ctx context.Context, userID, workspaceID primitive.ObjectID) error {
	// Get workspace and check permissions
	workspace, err := s.workspaceRepo.GetByID(ctx, workspaceID)
	if err != nil {
		return fmt.Errorf("failed to get workspace: %w", err)
	}

	// Only owner can delete workspace
	if workspace.OwnerID != userID {
		return fmt.Errorf("access denied: only workspace owner can delete workspace")
	}

	// Delete workspace
	_, err = s.workspaceRepo.Delete(ctx, workspaceID)
	if err != nil {
		return fmt.Errorf("failed to delete workspace: %w", err)
	}

	return nil
}

// List retrieves workspaces with pagination
func (s *WorkspaceService) List(ctx context.Context, userID primitive.ObjectID, limit, skip int64) ([]*WorkspaceResponse, int64, error) {
	// For now, just return workspaces where user is a member
	// In the future, this could be extended to include public workspaces
	workspaces, err := s.workspaceRepo.GetByUserID(ctx, userID)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to get workspaces: %w", err)
	}

	// Apply pagination manually for now
	total := int64(len(workspaces))
	start := skip
	end := skip + limit

	if start >= total {
		return []*WorkspaceResponse{}, total, nil
	}

	if end > total {
		end = total
	}

	paginatedWorkspaces := workspaces[start:end]
	responses := make([]*WorkspaceResponse, len(paginatedWorkspaces))
	for i, workspace := range paginatedWorkspaces {
		responses[i] = s.toWorkspaceResponse(workspace)
	}

	return responses, total, nil
}

// Member management requests and responses
type InviteMemberRequest struct {
	Email string `json:"email" validate:"required,email"`
	Role  string `json:"role" validate:"required,oneof=admin member"`
}

type UpdateMemberRoleRequest struct {
	Role string `json:"role" validate:"required,oneof=admin member"`
}

type MemberResponse struct {
	UserID   string    `json:"userId"`
	Email    string    `json:"email"`
	Name     string    `json:"name"`
	Role     string    `json:"role"`
	JoinedAt time.Time `json:"joinedAt"`
}

// InviteMember invites a user to the workspace
func (s *WorkspaceService) InviteMember(ctx context.Context, userID, workspaceID primitive.ObjectID, req InviteMemberRequest) error {
	// Get workspace and check permissions
	workspace, err := s.workspaceRepo.GetByID(ctx, workspaceID)
	if err != nil {
		return fmt.Errorf("failed to get workspace: %w", err)
	}

	// Check if user is admin
	if !workspace.IsAdmin(userID) {
		return fmt.Errorf("access denied: user is not an admin of this workspace")
	}

	// Validate role
	if req.Role != models.WorkspaceRoleAdmin && req.Role != models.WorkspaceRoleMember {
		return fmt.Errorf("invalid role: %s", req.Role)
	}

	// Check if user exists
	invitedUser, err := s.userRepo.GetByEmail(ctx, req.Email)
	if err != nil {
		return fmt.Errorf("user with email %s not found", req.Email)
	}

	// Check if user is already a member
	if workspace.IsMember(invitedUser.ID) {
		return fmt.Errorf("user is already a member of this workspace")
	}

	// Add member to workspace
	err = s.workspaceRepo.AddMember(ctx, workspaceID, invitedUser.ID, req.Role)
	if err != nil {
		return fmt.Errorf("failed to add member: %w", err)
	}

	// Send invitation email
	inviterUser, err := s.userRepo.GetByID(ctx, userID)
	if err != nil {
		return fmt.Errorf("failed to get inviter user: %w", err)
	}

	inviterName := inviterUser.FirstName + " " + inviterUser.LastName
	if inviterName == " " {
		inviterName = inviterUser.Email
	}

	subject := fmt.Sprintf("You've been invited to join %s", workspace.Name)
	body := fmt.Sprintf("Hi %s,\n\n%s has invited you to join the workspace '%s' as a %s.\n\nYou can access the workspace by logging into your account.\n\nBest regards,\nProject Management Team",
		invitedUser.FirstName, inviterName, workspace.Name, req.Role)

	err = s.emailService.SendEmail(req.Email, subject, body)
	if err != nil {
		// Log error but don't fail the operation
		fmt.Printf("Failed to send invitation email: %v\n", err)
	}

	return nil
}

// RemoveMember removes a member from the workspace
func (s *WorkspaceService) RemoveMember(ctx context.Context, userID, workspaceID, memberUserID primitive.ObjectID) error {
	// Get workspace and check permissions
	workspace, err := s.workspaceRepo.GetByID(ctx, workspaceID)
	if err != nil {
		return fmt.Errorf("failed to get workspace: %w", err)
	}

	// Check if user is admin or removing themselves
	if !workspace.IsAdmin(userID) && userID != memberUserID {
		return fmt.Errorf("access denied: user is not an admin of this workspace")
	}

	// Cannot remove workspace owner
	if workspace.OwnerID == memberUserID {
		return fmt.Errorf("cannot remove workspace owner")
	}

	// Remove member from workspace
	err = s.workspaceRepo.RemoveMember(ctx, workspaceID, memberUserID)
	if err != nil {
		return fmt.Errorf("failed to remove member: %w", err)
	}

	return nil
}

// UpdateMemberRole updates a member's role
func (s *WorkspaceService) UpdateMemberRole(ctx context.Context, userID, workspaceID, memberUserID primitive.ObjectID, req UpdateMemberRoleRequest) error {
	// Get workspace and check permissions
	workspace, err := s.workspaceRepo.GetByID(ctx, workspaceID)
	if err != nil {
		return fmt.Errorf("failed to get workspace: %w", err)
	}

	// Check if user is admin
	if !workspace.IsAdmin(userID) {
		return fmt.Errorf("access denied: user is not an admin of this workspace")
	}

	// Validate role
	if req.Role != models.WorkspaceRoleAdmin && req.Role != models.WorkspaceRoleMember {
		return fmt.Errorf("invalid role: %s", req.Role)
	}

	// Cannot change owner's role
	if workspace.OwnerID == memberUserID {
		return fmt.Errorf("cannot change workspace owner's role")
	}

	// Update member role
	err = s.workspaceRepo.UpdateMemberRole(ctx, workspaceID, memberUserID, req.Role)
	if err != nil {
		return fmt.Errorf("failed to update member role: %w", err)
	}

	return nil
}

// GetMembers retrieves all members of a workspace with user details
func (s *WorkspaceService) GetMembers(ctx context.Context, userID, workspaceID primitive.ObjectID) ([]*MemberResponse, error) {
	// Get workspace and check permissions
	workspace, err := s.workspaceRepo.GetByID(ctx, workspaceID)
	if err != nil {
		return nil, fmt.Errorf("failed to get workspace: %w", err)
	}

	// Check if user has access to this workspace
	if !workspace.IsMember(userID) {
		return nil, fmt.Errorf("access denied: user is not a member of this workspace")
	}

	// Get member details
	var members []*MemberResponse
	for _, member := range workspace.Members {
		user, err := s.userRepo.GetByID(ctx, member.UserID)
		if err != nil {
			// Skip members whose user records are not found
			continue
		}

		name := user.FirstName + " " + user.LastName
		if name == " " {
			name = user.Email
		}

		members = append(members, &MemberResponse{
			UserID:   member.UserID.Hex(),
			Email:    user.Email,
			Name:     name,
			Role:     member.Role,
			JoinedAt: member.JoinedAt,
		})
	}

	return members, nil
}

// LeaveWorkspace allows a user to leave a workspace
func (s *WorkspaceService) LeaveWorkspace(ctx context.Context, userID, workspaceID primitive.ObjectID) error {
	// Get workspace
	workspace, err := s.workspaceRepo.GetByID(ctx, workspaceID)
	if err != nil {
		return fmt.Errorf("failed to get workspace: %w", err)
	}

	// Cannot leave if user is the owner
	if workspace.OwnerID == userID {
		return fmt.Errorf("workspace owner cannot leave workspace. Transfer ownership or delete the workspace instead")
	}

	// Check if user is a member
	if !workspace.IsMember(userID) {
		return fmt.Errorf("user is not a member of this workspace")
	}

	// Remove user from workspace
	err = s.workspaceRepo.RemoveMember(ctx, workspaceID, userID)
	if err != nil {
		return fmt.Errorf("failed to leave workspace: %w", err)
	}

	return nil
}

// toWorkspaceResponse converts a workspace model to response
func (s *WorkspaceService) toWorkspaceResponse(workspace *models.Workspace) *WorkspaceResponse {
	return &WorkspaceResponse{
		ID:          workspace.ID.Hex(),
		Name:        workspace.Name,
		Description: workspace.Description,
		Logo:        workspace.Logo,
		OwnerID:     workspace.OwnerID.Hex(),
		Members:     workspace.Members,
		Settings:    workspace.Settings,
		CreatedAt:   workspace.CreatedAt,
		UpdatedAt:   workspace.UpdatedAt,
	}
}

// WorkspaceServiceInterface defines the contract for workspace services
type WorkspaceServiceInterface interface {
	Create(ctx context.Context, userID primitive.ObjectID, req CreateWorkspaceRequest) (*WorkspaceResponse, error)
	GetByID(ctx context.Context, userID, workspaceID primitive.ObjectID) (*WorkspaceResponse, error)
	GetByUserID(ctx context.Context, userID primitive.ObjectID) ([]*WorkspaceResponse, error)
	Update(ctx context.Context, userID, workspaceID primitive.ObjectID, req UpdateWorkspaceRequest) (*WorkspaceResponse, error)
	Delete(ctx context.Context, userID, workspaceID primitive.ObjectID) error
	List(ctx context.Context, userID primitive.ObjectID, limit, skip int64) ([]*WorkspaceResponse, int64, error)
	InviteMember(ctx context.Context, userID, workspaceID primitive.ObjectID, req InviteMemberRequest) error
	RemoveMember(ctx context.Context, userID, workspaceID, memberUserID primitive.ObjectID) error
	UpdateMemberRole(ctx context.Context, userID, workspaceID, memberUserID primitive.ObjectID, req UpdateMemberRoleRequest) error
	GetMembers(ctx context.Context, userID, workspaceID primitive.ObjectID) ([]*MemberResponse, error)
	LeaveWorkspace(ctx context.Context, userID, workspaceID primitive.ObjectID) error
}
